### PR TITLE
bugfix #130

### PR DESCRIPTION
**Pull request by Wikunia**
Don't call objective value if status isn't optimal.

### CODE REVIEW

**Comment by ccoffrin** [comment]
somewhat off topic, but you might want to add, `ALMOST_OPTIMAL` to `state_is_optimal`

**Comment by Wikunia** [comment]
> somewhat off topic, but you might want to add, `ALMOST_OPTIMAL` to `state_is_optimal`

thanks I forgot that the default is false :/

**Comment by ccoffrin** [comment]
And in this line, https://github.com/lanl-ansi/Juniper.jl/blob/512ff3814f4fba0d5f39da8fed8c1a2ffece2ddb/src/util.jl#L92

You should add `MOI.ALMOST_OPTIMAL` if you want to have the same semantics of `allow_almost` across NLP solvers are other convex solvers, like Gurobi.

**Comment by ccoffrin** [comment]
LGTM.  Make sure to add a note to the change log.

@odow, would you like to test your specific use case is working?

**Comment by odow** [comment]
Now I get a different error
```Julia
julia> using JuMP, Juniper, Gurobi, Ipopt

julia> const env = Gurobi.Env()
Academic license - for non-commercial use only
Gurobi.Env(Ptr{Nothing} @0x00007fe7ae281a00)

julia> model = Model(with_optimizer(
           Juniper.Optimizer, 
           Dict{Symbol, Any}(
               :nl_solver => with_optimizer(Ipopt.Optimizer, print_level = 0),
               :mip_solver => with_optimizer(Gurobi.Optimizer, env, OutputFlag=0)
           )
       ))
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Juniper

julia> @variable(model, 1 <= x[1:11], Int)
11-element Array{VariableRef,1}:
 x[1] 
 x[2] 
 x[3] 
 x[4] 
 x[5] 
 x[6] 
 x[7] 
 x[8] 
 x[9] 
 x[10]
 x[11]

julia> @constraint(model, [i=2:11], x[i-1] <= x[i] - 1)
1-dimensional DenseAxisArray{ConstraintRef{Model,C,Shape} where Shape<:AbstractShape where C,1,...} with index sets:
    Dimension 1, 2:11
And data, a 10-element Array{ConstraintRef{Model,C,Shape} where Shape<:AbstractShape where C,1}:
 x[1] - x[2] ≤ -1.0  
 x[2] - x[3] ≤ -1.0  
 x[3] - x[4] ≤ -1.0  
 x[4] - x[5] ≤ -1.0  
 x[5] - x[6] ≤ -1.0  
 x[6] - x[7] ≤ -1.0  
 x[7] - x[8] ≤ -1.0  
 x[8] - x[9] ≤ -1.0  
 x[9] - x[10] ≤ -1.0 
 x[10] - x[11] ≤ -1.0

julia> @NLconstraint(model, sum(1 / x[i] for i in 1:11) == 2)
(1.0 / x[1] + 1.0 / x[2] + 1.0 / x[3] + 1.0 / x[4] + 1.0 / x[5] + 1.0 / x[6] + 1.0 / x[7] + 1.0 / x[8] + 1.0 / x[9] + 1.0 / x[10] + 1.0 / x[11]) - 2.0 = 0

julia> optimize!(model)
nl_solver   : OptimizerFactory(Ipopt.Optimizer, (), Base.Iterators.Pairs(:print_level=>0))
mip_solver  : OptimizerFactory(Gurobi.Optimizer, (Gurobi.Env(Ptr{Nothing} @0x00007fe7ae281a00),), Base.Iterators.Pairs(:OutputFlag=>0))
log_levels  : Symbol[:Options, :Table, :Info]

#Variables: 11
#IntBinVar: 11
#Constraints: 11
#Linear Constraints: 10
#Quadratic Constraints: 0
#NonLinear Constraints: 1
Obj Sense: Max


******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit http://projects.coin-or.org/Ipopt
******************************************************************************

Status of relaxation: LOCALLY_SOLVED
Time for relaxation: 6.911518096923828
Relaxation Obj: 0.0

       MIPobj              NLPobj       Time 
=============================================
       2.9268              0.0128        5.2 
... lines omitted ...
      13.1459             15.7105        7.8 
┌ Warning: MIP couldn't be solved to optimality. Terminated with status: INFEASIBLE
└ @ Juniper ~/.julia/packages/Juniper/5Cx7l/src/fpump.jl:337

FP: 7.87506103515625 s
FP: 130 rounds
FP: No integral solution found

 ONodes   CLevel          Incumbent                   BestBound            Gap    Time   Restarts  GainGap  
============================================================================================================
┌ Warning: Changed num_strong_var to 9 because of strong_branching_approx_time_limit
└ @ Juniper ~/.julia/packages/Juniper/5Cx7l/src/bb_strategies.jl:261
    2       2                 -                          0.0                -     1.7       0         -     
    3       3                 -                          0.0                -     2.2       -        0.0%   
    4       4                 -                          0.0                -     2.2       -        0.0%   
    5       5                 -                          0.0                -     2.2       -        0.0%   
    6       6                 -                          0.0                -     2.2       -        0.0%   
    7       7                 -                          0.0                -     2.3       -        0.0%   
    7       8                 -                          0.0                -     2.3       -        0.0%   
    8       9                 -                          0.0                -     2.3       -        0.0%   
    9       10                -                          0.0                -     2.3       -        0.0%   
   10       11                -                          0.0                -     2.3       -        0.0%   
   10       12                -                          0.0                -     2.3       -        0.0%   
    9       13                -                          0.0                -     2.4       -        0.0%   
   10       12                -                          0.0                -     2.4       -        0.0%   
   10       13                -                          0.0                -     2.4       -        0.0%   
   10       14                -                          0.0                -     2.4       -        0.0%   
   11       15                -                          0.0                -     2.5       -        0.0%   
   12       16                -                          0.0                -     2.5       -        0.0%   
   13       17                -                          0.0                -     2.5       -        0.0%   
   14       18                -                          0.0                -     2.5       -        0.0%   
   15       19                -                          0.0                -     2.5       -        0.0%   
   16       20                -                          0.0                -     2.5       -        0.0%   
   17       21                -                          0.0                -     2.5       -        0.0%   
   18       22                -                          0.0                -     2.6       -        0.0%   
   19       23                -                          0.0                -     2.6       -        0.0%   
   20       24                -                          0.0                -     2.6       -        0.0%   
┌ Warning: Only almost locally solved
└ @ Juniper ~/.julia/packages/Juniper/5Cx7l/src/BnBTree.jl:89
   21       25                -                          0.0                -     2.6       -        0.0%   
   22       3                 -                          NaN                -     2.6       -        0.0%   
   22       4                 -                          NaN                -     2.6       -        0.0%   
   23       5                 -                          NaN                -     2.7       -        0.0%   
   24       6                 -                          NaN                -     2.7       -        0.0%   
   25       7                 -                          NaN                -     2.7       -        0.0%   
   26       8                 -                          NaN                -     2.7       -        0.0%   
   27       10                -                          NaN                -     2.7       -        0.0%   
   28       11                -                          NaN                -     2.7       -        0.0%   
   28       13                -                          NaN                -     2.7       -        0.0%   
   29       16                -                          NaN                -     2.7       -        0.0%   
   30       17                -                          NaN                -     2.8       -        0.0%   
   31       18                -                          NaN                -     2.8       -        0.0%   
   32       19                -                          NaN                -     2.8       -        0.0%   
   33       20                -                          NaN                -     2.8       -        0.0%   
   34       21                -                          NaN                -     2.8       -        0.0%   
   35       22                -                          NaN                -     2.8       -        0.0%   
   36       23                -                          NaN                -     2.8       -        0.0%   
   37       24                -                          NaN                -     2.8       -        0.0%   
   38       25                -                          NaN                -     2.8       -        0.0%   
ERROR: AssertionError: !(isapprox(u_b, l_b; atol=atol))
Stacktrace:
 [1] branch_pseudo(::Juniper.JuniperProblem, ::Juniper.BnBNode, ::Array{Int64,1}, ::Juniper.GainObj, ::Float64, ::Float64) at /Users/oscar/.julia/packages/Juniper/5Cx7l/src/bb_strategies.jl:368
 [2] get_branching_disc_idx!(::Juniper.JuniperProblem, ::Juniper.StepObj, ::Juniper.SolverOptions, ::Array{Int64,1}, ::Juniper.GainObj, ::Int64) at /Users/oscar/.julia/packages/Juniper/5Cx7l/src/BnBTree.jl:32
 [3] one_branch_step!(::Juniper.JuniperProblem, ::Nothing, ::Juniper.SolverOptions, ::Juniper.StepObj, ::Array{Int64,1}, ::Juniper.GainObj, ::Int64) at /Users/oscar/.julia/packages/Juniper/5Cx7l/src/BnBTree.jl:312
 [4] solve_sequential(::Juniper.BnBTreeObj, ::Array{Any,1}, ::Float64, ::Array{Symbol,1}, ::Array{Int64,1}, ::Juniper.TimeObj) at /Users/oscar/.julia/packages/Juniper/5Cx7l/src/BnBTree.jl:408
 [5] solvemip(::Juniper.BnBTreeObj) at /Users/oscar/.julia/packages/Juniper/5Cx7l/src/BnBTree.jl:603
 [6] optimize!(::Juniper.Optimizer) at /Users/oscar/.julia/packages/Juniper/5Cx7l/src/MOI_wrapper/MOI_wrapper.jl:293
 [7] optimize!(::MathOptInterface.Bridges.LazyBridgeOptimizer{Juniper.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}) at /Users/oscar/.julia/packages/MathOptInterface/C3lip/src/Bridges/bridgeoptimizer.jl:73
 [8] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}) at /Users/oscar/.julia/packages/MathOptInterface/C3lip/src/Utilities/cachingoptimizer.jl:170
 [9] #optimize!#79(::Bool, ::Bool, ::Function, ::Model, ::Nothing) at /Users/oscar/.julia/packages/JuMP/tyMag/src/optimizer_interface.jl:132
 [10] optimize! at /Users/oscar/.julia/packages/JuMP/tyMag/src/optimizer_interface.jl:105 [inlined] (repeats 2 times)
 [11] top-level scope at none:0

(v1.1) pkg> st
    Status `~/.julia/environments/v1.1/Project.toml`
  [6e4b80f9] BenchmarkTools v0.4.2
  [60bf3e95] GLPK v0.10.0
  [2e9cd046] Gurobi v0.6.0
  [b6b21f68] Ipopt v0.5.4
  [682c06a0] JSON v0.20.0
  [4076af6c] JuMP v0.19.1
  [2ddba703] Juniper v0.4.1 #bugfix-gurobi-130 (https://github.com/lanl-ansi/Juniper.jl.git)
  [f4570300] MathOptFormat v0.0.0 #master (https://github.com/odow/MathOptFormat.jl)
  [91a5bcdd] Plots v0.25.1
  [fcd29c91] QuantEcon v0.16.0
  [c946c3f1] SCS v0.5.1
  [f4570300] SDDP v0.0.0 [`~/.julia/dev/SDDP`]
```

**Comment by Wikunia** [comment]
Thanks @odow for testing again. I missed one `allow_almost`. I've added the bugfix and this test case as it's reasonably fast and has a lot of almost solved cases.
Further more I think it's reasonable to return the status `ALMOST_LOCALLY_SOLVED` in those cases instead of `LOCALLY_SOLVED`. Correct @ccoffrin ?
Mentioned both in the change log fo v0.4.2

**Comment by ccoffrin** [comment]
Regarding the termination status of Juniper.  I think it should match the status that the subsolver had when it found the incumbent solution.  In that case, we might consider revising the incumbent update logic; if your current incumbent is `ALMOST_LOCALLY_SOLVED` and find a solution that has the same objective value (up to some tolerance) but the status is `LOCALLY_SOLVED`, then it should use the `LOCALLY_SOLVED` solution for the incumbent.

Juniper should also have a configuration flag for the user to specify if `ALMOST_LOCALLY_SOLVED` is acceptable or not; with acceptable being the default value.

In the code I see a lot of `ALMOST_LOCALLY_SOLVED` and `LOCALLY_SOLVED`, it would probably be best to abstract these into some functions helper functions, so that we can easily support solvers that report other reasonable status like `OPTIMAL`,`ALMOST_OPTIMAL`.  I could imagine someone might want to use a global optimization solver in place of Ipopt, which would return the `OPTIMAL` status values.

**Comment by Wikunia** [comment]
One question about the solver option. You mean only allowing `LOCALLY_SOLVED` or `OPTIMAL` in all sub problems and if it's only almost solved disregard it as `Infeasible`? 
There is a solver option for disable `ALMOST` in integral solutions with the default being to allow it and throwing a warning.

**Comment by ccoffrin** [comment]
You make a good point.  I think one could make the case for both options.  Possibly the Juniper solver settings should allow for three cases,
- Treat `LOCALLY_X` and `X` the same
- `LOCALLY_X` and `X` are fine during search but only consider `X` for the incumbent
- Treat `LOCALLY_X` as infeasible / numeric error

File: src/fpump.jl [hunk 1]
```diff
@@ -390,16 +385,17 @@ function fpump(optimizer, m)
...
+            if state_is_optimal(real_status) || (only_almost_solved(real_status) && m.options.allow_almost_solved_integral)
+                if only_almost_solved(real_status)
+                    @warn "Integral feasible point only almost solved. Disable with `allow_almost_solved_integral=false`"
                 end 
                 nlp_obj = real_obj
                 nlp_sol = real_sol
                 iscorrect = true
                 break
             elseif are_type_correct(nlp_sol, m.var_type, m.disc2var_idx, catol)
+                # TODO: maybe there is a different better status code
```
Comment (Wikunia) [inline src/fpump.jl]: @ccoffrin don't know what the best status is in this situation. 
Here the feasibility pump was able to find an integral solution but optimizing it (changing the real valued variables) wasn't successful. In that case the values from the nlp are just plugged into the objective function to get an objective value.

Comment (ccoffrin) [reply src/fpump.jl]: I would say `LOCALLY_SOLVED`/`FEASIBLE_POINT` is the status there.  Just make sure to throw a warning.

**Comment by Wikunia** [comment]
> * Treat `LOCALLY_X` and `X` the same
> * `LOCALLY_X` and `X` are fine during search but only consider `X` for the incumbent
> * Treat `LOCALLY_X` as infeasible / numeric error

Alright will be the next step. Additionally currently when setting `allow_almost_solved_integral` to false. I do a restart once and hope that it returns `SOLVED` next time. This should probably also be an option.

File: src/fpump.jl (hunk 1 above)
Comment (Wikunia) [reply src/fpump.jl]: Does this depend on the solver? Is it possible to call it OPTIMAL if the nl solver is globally optimal?
Warning is thrown ;)

Comment (ccoffrin) [reply src/fpump.jl]: Nothing that comes out of the feasibility pump can be called `OPTIMAL` regardless of the solver.

Comment (Wikunia) [reply src/fpump.jl]: haha thanks that's a very valid point totally forgot that this is all in the feasibility pump.

**Comment by ccoffrin** [comment]
@Wikunia, shall we merge this?  Or were you still working on it?

**Comment by Wikunia** [comment]
I'm currently rewriting the incumbent stuff to have a cleaner representation of status codes and will add the solver options for allowing/disable almost solved status codes. 
If merging isn't urgent I would add this to this branch and merge afterwards

**Comment by Wikunia** [comment]
Can be merged if it looks good to you @ccoffrin

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,13 @@
 # Juniper.jl Changelog
 
+### v0.4.2
+- Ability to not allowing almost solved status codes in tree search
```
Comment (ccoffrin) [inline CHANGELOG.md]: "not accept"

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,7 +1,13 @@
...
+- Better handling of status codes internally
```
Comment (ccoffrin) [inline CHANGELOG.md]: "MOI solver status codes"

**Pull request merged**